src/vehicle_autodrive.cpp

```diff
@@ -0,0 +1,1046 @@
+    }
+};
+
+/**
+ * The address of a navigation node, i.e. a position and orientation on the nav map.
+ */
+// NOLINTNEXTLINE(cata-xy)
+/*
+ * A node address annotated with its heuristic score, an approximation of how
+ * much it would cost to reach the goal from this node.
+ */
+struct scored_address {
+    struct node_address addr;
+    int16_t score;
+    bool operator> ( const scored_address &other ) const {
+        return score >= other.score;
+    }
+};
+
+/*
+ * Data structure representing a navigation node that is known to be reachable. Contains
+ * information about the path to get there and enough information about the state of
+ * the vehicle and driver to predict which nodes can be reached on the next turn.
+ */
+struct navigation_node {
+    node_address prev;
+    int16_t cost;
+    int16_t tileray_steps;
+    int16_t speed;
+    int16_t target_speed_tps;
+};
+
+enum class collision_check_result : int {
+    ok,
+    no_visibility,
+    close_obstacle,
+    slow_down
+};
+
+class vehicle::autodrive_controller
+{
+    public:
+        explicit autodrive_controller( const vehicle &driven_veh, const Character &driver );
+        const Character &get_driver() {
+            return driver;
+        }
+        const auto_navigation_data &get_data() {
+            return data;
+        }
+        void check_safe_speed();
+        cata::optional<navigation_step> compute_next_step();
+        collision_check_result check_collision_zone( orientation turn_dir );
+        void reduce_speed();
+
+    private:
+        const vehicle &driven_veh;
+        const Character &driver;
+        auto_navigation_data data;
+
+        bool check_drivable( tripoint pt ) const;
+        void compute_obstacles();
+        vehicle_profile compute_profile( orientation facing ) const;
+        void compute_valid_positions();
+        void precompute_data();
+        scored_address compute_node_score( const node_address &addr, const navigation_node &node ) const;
+        void compute_next_nodes( const node_address &addr, const navigation_node &node,
+                                 int target_speed_tps,
+                                 std::vector<std::pair<node_address, navigation_node>> &next_nodes ) const;
+        cata::optional<std::vector<navigation_step>> compute_path( int speed_tps ) const;
+};
+
+
+static const std::array<orientation, NUM_ORIENTATIONS> &all_orientations()
+{
+    static const auto orientations_array = [] {
+        std::array<orientation, NUM_ORIENTATIONS> ret;
+        for( int i = 0; i < NUM_ORIENTATIONS; i++ )
+        {
+            ret[i] = static_cast<orientation>( i );
+        }
+        return ret;
+    }();
+    return orientations_array;
+}
+
+/*
+ * Normalize an orientation value to the range [0, NUM_ORIENTATIONS-1] using modular arithmetic.
+ */
+static orientation normalize_orientation( int angle )
+{
+    return static_cast<orientation>( angle % NUM_ORIENTATIONS + ( angle < 0 ? NUM_ORIENTATIONS : 0 ) );
```

Comment:
A branch-less way to do this:
```suggestion
    return static_cast<orientation>( ( angle + NUM_ORIENTATIONS ) % NUM_ORIENTATIONS );
```

Reply:
You're the second person to suggest it, but that doesn't work for angle < NUM_ORIENTATIONS, since the % operator returns negative numbers when the first operand is negative. You can make it branchless like this:
( angle % NUM_ORIENTATIONS + NUM_ORIENTATIONS ) % NUM_ORIENTATIONS

I guess I'll change it to that and add a warning comment.

Reply:
🤣 I didn't think about `angle` < -`NUM_ORIENTATIONS`

Reply:
I was the first person to suggest it, BTW.

Reply:
A second modulus operation is probably [much slower than a branch](https://quick-bench.com/q/oJbVwLr9G5HJb0oRaYpQOCec4E4) (on most platforms it won't actually be a branch; just a conditional move) so I think the original implementation is probably faster.  Though personally, I'd do something like this for code clarity, because trying to squeeze it all in one line is a little overwhelming:
```
angle %= NUM_ORIENTATIONS;
if( angle < 0 ) {
  angle += NUM_ORIENTATIONS;
}
return static_cast<orientation>( angle );
```
(Sorry, this is probably too much bikeshedding on something that really doesn't matter that much...)

Reply:
FWIW, I realised that QuickBench I linked (taken from someone else's StackOverflow answer) is not really representative of the situation at hand.  [Here's a better one](https://quick-bench.com/q/04WpfD9LrgZd_D1fJQSfHHB_ass) with a similar conclusion.

Reply:
Neat benchmark!
Well, this won't make any difference in performance but sure, I'll go with that suggestion.